src/finufft.cpp

```diff
@@ -1112,26 +1128,110 @@ int FINUFFT_EXECUTE(FINUFFT_PLAN p, CPX *cj, CPX *fk) {
+              // do only parts of axis 1
+              ducc0::c2c(sub1, sub1, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub2, sub2, {1}, p->fftSign < 0, FLT(1), nthreads);
+              // do axis 2 in full
+              ducc0::c2c(data, data, {2}, p->fftSign < 0, FLT(1), nthreads);
+            }
+          }
+        } else {                          // 3D
+          if ((p->ms < 2) || (p->mt < 2)) // something is weird, do standard FFT
+            ducc0::c2c(data, data, axes, p->fftSign < 0, FLT(1), nthreads);
+          else {
+            size_t z_lo = size_t((p->ms + 1) / 2);
+            size_t z_hi = size_t(ns[2] - p->ms / 2);
+            size_t y_lo = size_t((p->mt + 1) / 2);
+            size_t y_hi = size_t(ns[1] - p->mt / 2);
+            auto sub1   = ducc0::subarray(data, {{}, {}, {}, {0, z_lo}});
+            auto sub2   = ducc0::subarray(data, {{}, {}, {}, {z_hi, ducc0::MAXIDX}});
+            auto sub3   = ducc0::subarray(sub1, {{}, {}, {0, y_lo}, {}});
+            auto sub4   = ducc0::subarray(sub1, {{}, {}, {y_hi, ducc0::MAXIDX}, {}});
+            auto sub5   = ducc0::subarray(sub2, {{}, {}, {0, y_lo}, {}});
+            auto sub6   = ducc0::subarray(sub2, {{}, {}, {y_hi, ducc0::MAXIDX}, {}});
+            if (p->type == 1) { // spreading, not all parts of the output array are needed
+              // do axis 3 in full
+              ducc0::c2c(data, data, {3}, p->fftSign < 0, FLT(1), nthreads);
+              // do only parts of axis 2
+              ducc0::c2c(sub1, sub1, {2}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub2, sub2, {2}, p->fftSign < 0, FLT(1), nthreads);
+              // do even smaller parts of axis 1
+              ducc0::c2c(sub3, sub3, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub4, sub4, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub5, sub5, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub6, sub6, {1}, p->fftSign < 0, FLT(1), nthreads);
+            } else { // interpolation, parts of the input array are zero
+              // do even smaller parts of axis 1
+              ducc0::c2c(sub3, sub3, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub4, sub4, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub5, sub5, {1}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub6, sub6, {1}, p->fftSign < 0, FLT(1), nthreads);
+              // do only parts of axis 2
+              ducc0::c2c(sub1, sub1, {2}, p->fftSign < 0, FLT(1), nthreads);
+              ducc0::c2c(sub2, sub2, {2}, p->fftSign < 0, FLT(1), nthreads);
+              // do axis 3 in full
+              ducc0::c2c(data, data, {3}, p->fftSign < 0, FLT(1), nthreads);
+            }
+          }
+        }
+        delete[] ns;
+      }
+#else
       FFTW_EX(p->fftwPlan); // if thisBatchSize<batchSize it wastes some flops
+#endif
```

Comment:
can this be moved to fft.cpp ?

Reply:
I'll give it a try, but in my eyes that just makes the whole thing harder to read, since there is no real clear dividing line between spread/interp and FFT.